sstables/sstable_set.hh

```diff
@@ -47,24 +47,86 @@ public:
 // Structure holds all sstables (a.k.a. fragments) that belong to same run identifier, which is an UUID.
 // SStables in that same run will not overlap with one another.
 class sstable_run {
-    sstable_list _all;
+    std::unordered_set<shared_sstable> _all;
 public:
+    sstable_run(std::unordered_set<shared_sstable> all) : _all(std::move(all)) {}
     void insert(shared_sstable sst);
     void erase(shared_sstable sst);
     // Data size of the whole run, meaning it's a sum of the data size of all its fragments.
     uint64_t data_size() const;
-    const sstable_list& all() const { return _all; }
+    const std::unordered_set<shared_sstable>& all() const { return _all; }
 };
 
-class sstable_set {
+class versioned_sstable_set_data {
     std::unique_ptr<sstable_set_impl> _impl;
-    schema_ptr _schema;
+    std::unordered_map<utils::UUID, std::unordered_set<shared_sstable>> _all_runs;
+    std::unordered_map<shared_sstable, unsigned> _times_added;
+    std::set<shared_sstable> _all;
+public:
+    versioned_sstable_set_data(std::unique_ptr<sstable_set_impl> impl);
+    void insert(shared_sstable sst);
+    void erase(shared_sstable sst);
+    std::vector<shared_sstable> select(const dht::partition_range& range) const;
+    std::unordered_set<shared_sstable> select_by_run_id(utils::UUID run_id) const;
+    const std::set<shared_sstable>& all() const;
+    void remove(shared_sstable sst);
+    std::unique_ptr<incremental_selector_impl> make_incremental_selector();
+};
+
+class sstable_set_version;
+
+class sstable_set_version_reference {
```

Comment:
Can't this class be folded into `sstable_list` or replaced with an `lw_shared_ptr<>`?

Reply:
It can't be lw_shared_ptr because it is a reference to a sstable_set_version, and a sstable_set_version has a member of this type and lw_shared_ptrs don't work on incomplete (so recursive too) types(at least I couldn't make them work), it could be an std::shared_ptr though, but it would still require more additional work than writing own pointer class. It could be folded into sstable_list but it would have to be folded into sstable_set_version at the same time, because both have a member of this type, and what is being done in its destructor would have to be copied to both of these classes' destructors too.

Reply:
There is some support for incomplete classes with lw_shared_ptr, see shared_sstable.hh, lw_shared_ptr_deleted<shared_sstable> for an example.

If we can keep sstable_set and lw_shared_ptr<sstable_set> naming (but now they refer to a version) that may eliminate a large amount of churn. But don't go overboard trying to make it work, if it doesn't, then it doesn't.